lottery/calypso/mod.go

```diff
@@ -0,0 +1,238 @@
+package calypso
+
+import (
+	"crypto/rand"
+	"encoding/json"
+
+	"go.dedis.ch/dela/serde"
+	serdej "go.dedis.ch/dela/serde/json"
+
+	"go.dedis.ch/dela/dkg"
+	"go.dedis.ch/dela/ledger/arc"
+	"go.dedis.ch/dela/ledger/arc/darc"
+	"go.dedis.ch/dela/lottery"
+	"go.dedis.ch/dela/lottery/storage"
+	"go.dedis.ch/dela/lottery/storage/inmemory"
+	"go.dedis.ch/dela/mino"
+	"go.dedis.ch/kyber/v3"
+	"go.dedis.ch/kyber/v3/suites"
+	"golang.org/x/xerrors"
+)
+
+const (
+	keySize = 32
+
+	// ArcRuleUpdate defines the rule to update the arc. This rule must be set
+	// at the write creation to allow the arc to be latter updated.
+	ArcRuleUpdate = "calypso_update"
+	// ArcRuleRead defines the arc rule to read a value
+	ArcRuleRead = "calypso_read"
+)
+
+// Suite is the Kyber suite for Pedersen.
+var suite = suites.MustFind("Ed25519")
+
+// NewCalypso creates a new Calypso
+func NewCalypso(dkg dkg.DKG) *Calypso {
+	return &Calypso{
+		dkg:        dkg,
+		storage:    inmemory.NewInMemory(),
+		serializer: serdej.NewSerializer(),
+	}
+}
+
+// Calypso is a wrapper around DKG to provide a storage and authorization layer
+//
+// implements lottery.Secret
+type Calypso struct {
+	dkg        dkg.DKG
+	dkgActor   dkg.Actor
+	storage    storage.KeyValue
+	serializer serde.Serializer
+}
+
+// Setup implements lottery.Secret
+func (c *Calypso) Setup(players mino.Players, pubKeys []kyber.Point,
+	threshold int) (pubKey kyber.Point, err error) {
+
+	actor, err := c.dkg.Listen()
+	if err != nil {
+		return nil, xerrors.Errorf("failed to listen dkg: %v", err)
+	}
+
+	pubKey, err = actor.Setup(players, pubKeys, threshold)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to setup dkg: %v", err)
+	}
+
+	c.dkgActor = actor
+
+	return pubKey, nil
+}
+
+// Write implements lottery.Secret
+func (c *Calypso) Write(message lottery.EncryptedMessage, ac arc.AccessControl) ([]byte, error) {
+	key := make([]byte, keySize)
+	_, err := rand.Read(key)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to generate random key: %v", err)
+	}
+
+	darcKey := make([]byte, keySize)
+	_, err = rand.Read(darcKey)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to generate random darcKey: %v", err)
+	}
+
+	kBuf, err := message.GetK().MarshalBinary()
+	if err != nil {
+		return nil, xerrors.Errorf("failed to marshal K: %v", err)
+	}
+
+	cBuf, err := message.GetC().MarshalBinary()
+	if err != nil {
+		return nil, xerrors.Errorf("failed to marshal C: %v", err)
+	}
+
+	acBuf, err := c.serializer.Serialize(ac)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to encode darc: %v", err)
+	}
+
+	messageJSON := encryptedJSON{
+		K:       kBuf,
+		C:       cBuf,
+		DarcKey: darcKey,
+	}
+
+	messageBuf, err := json.Marshal(messageJSON)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to marshal message: %v", err)
+	}
+
+	c.storage.Store(key, messageBuf)
```

Comment:
Why don't you use the hash to detect duplicates ? Also the access control should be stored alongside with the encrypted message.

Reply:
you mean using `H(K || C)` for the key instead of using a random one?

Reply:
Yes, and now that I think of it, you can't update the access control if we don't have a ledger.